ci: Add ecp-data-vis-sdk CI pipeline

chuckatkins opened the pull request.

chuckatkins commented:
Sorry for the churn, bear with me as I iterate on trying to get a better syntax in the `gitlab-ci.yml` to share common job information.

scottwittenburg commented:
> Sorry for the churn, bear with me as I iterate on trying to get a better syntax in the `gitlab-ci.yml` to share common job information.

Don't apologize!  I appreciate you churning on it, as will anyone who comes along after you attempting to add another stack.

scottwittenburg reviewed (commented):
This is looking great @chuckatkins, thanks for working to consolidate the bits in the `gitlab-ci.yml` 🎉 

Before we merge this, we need to cut the stack down to something smaller (to be discussed with @tgamblin on Friday), and make sure all the packages build with binary relocation testing enabled.

chuckatkins commented:
I'm seeing an issue where some jobs in the pipeline have build failures but the job still passes.  This necessarily causes problems in downstream dependents.  In particular you can see this with the sz package: https://gitlab.next.spack.io/spack/spack/-/jobs/191112 .  Here the job passes and it's dependents proceed but the log shows a build failure.

scottwittenburg commented:
Yes, I see that in your PR pipeline too, once you pointed it out. I was expecting the buildcache creation to fail when it couldn't find the supposedly installed hash, but maybe something I did recently changed that behavior.  I've been seeing failures to build specs result in pipeline failure, so I'm not sure what's going on here, but I'll look into it.

scottwittenburg commented on share/spack/gitlab/cloud_e4s_pipelines/stacks/data-vis-sdk/spack.yaml:
```diff
@@ -0,0 +1,69 @@
+        - spack --version
+        - cd share/spack/gitlab/cloud_e4s_pipelines/stacks/data-vis-sdk
+        - spack env activate --without-view .
+      tags: ["spack", "public", "medium", "x86_64"]
```
This section does not inherit anything from the root level of the `gitlab-ci` section in the same way the entries in `mappings` do, so `service-job-attributes` needs an image.  This is what is causing the failures of the last two jobs (`cleanup` and `rebuild-index`) with the error:

```
==> Error: SPACK_ROOT must point to spack's prefix when using sh
Run this with the correct prefix before sourcing setup-env.sh:
    export SPACK_ROOT=</path/to/spack>
```

tldahlgren commented:
Closing and re-opening to force re-running checks after nearly all failed.

chuckatkins commented:
Just adding "don't merge yet" so we can discuss with @tgamblin on our Friday call before merging.

chuckatkins commented:
@scottwittenburg looks like `sz+python` is failing like it should now.

scottwittenburg commented:
> @scottwittenburg looks like `sz+python` is failing like it should now.

Good to hear, thanks for checking 😅

tgamblin commented:
Per our discussion, I've converted this to draft until @scottwittenburg comes back from break -- we want to make sure pipeline folks are around before we merge to develop.

tldahlgren commented:
Closing and re-opening to force what appear to be stale checks to run.

Commit:
ci: Enable an initial subset of the ECP Data & Vis SDK CI pipeline

scottwittenburg commented on share/spack/gitlab/cloud_pipelines/stacks/data-vis-sdk/spack.yaml:
```diff
@@ -0,0 +1,52 @@
+    script:
+      - . "./share/spack/setup-env.sh"
+      - spack --version
+      - cd share/spack/gitlab/cloud_pipelines/stacks/data-vis-sdk
```
This now needs to activate the concretized environment passed as artifact from the pipeline generation job.  So:

```
cd ${SPACK_CONCRETE_ENV_DIR}
```

Similar to the E4S one [here](https://github.com/spack/spack/blob/develop/share/spack/gitlab/cloud_pipelines/stacks/e4s/spack.yaml#L124).

scottwittenburg replied:
After you rebase, that is...

scottwittenburg replied:
Then I can empty your PR mirror if we want to see everything rebuild again.

chuckatkins commented:
@scottwittenburg Finally! Good to go!

scottwittenburg commented on share/spack/gitlab/cloud_pipelines/stacks/data-vis-sdk/spack.yaml:
```diff
@@ -0,0 +1,68 @@
+    all:
+      target: [x86_64]
+
+  # Currently only hdf5 is inebaled to reduce the load on the CI build
```
I don't think this comment is true anymore, is it?

Commit:
dvsdk: Turn off variants by default

This allows an install to more easily be explicit about which pieces to
turn on as more variants are added

Commit:
dvsdk: effectively disable the broken variants

chuckatkins commented:
🎉🎉🎉🎉🎉🎉🎉🎉🎉

tgamblin commented:
Nice!
